File: Finished/valid_parentheses.java
import java.util.*;

class Solution {
  public static char getComplementBracket(char bracket) {
    if (bracket == '(') {
      return ')';
    } else if (bracket == '[') {
      return ']';
    } else if (bracket == '{') {
      return '}';
    } else {
      return ' ';
    }
  }

  public static boolean isClosingBracket(char bracket) {
    return bracket == ')' || bracket == ']' || bracket == '}';
  }

  public boolean isValid(String s) {
    Stack<Character> stack = new Stack<>();

    for (int i = 0; i < s.length(); i++) {
      char bracket = s.charAt(i);

      if (isClosingBracket(bracket)) {
        if (stack.isEmpty()) {
          return false;
        }

        char top = stack.pop();

        if (getComplementBracket(top) != bracket) {
          return false;
        }
      } else {
        stack.push(bracket);
      }
    }
    return stack.isEmpty();
  }
}

File: Finished/validParentheses.js
// https://leetcode.com/problems/valid-parentheses/

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  function getComp(bracket) {
    switch (bracket) {
      case "(":
        return ")";
      case "[":
        return "]";
      case "{":
        return "}";
      default:
        return "";
    }
  }

  function isOpeningBracket(s) {
    return s === "(" || s === "[" || s === "{";
  }

  const stack = [];

  for (const b of s) {
    if (isOpeningBracket(b)) {
      stack.push(b);
      continue;
    }

    if (stack.length === 0) return false;

    const top = stack.pop();

    if (b !== getComp(top)) return false;
  }

  return stack.length === 0;
};
